samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaSystemProducer.scala

```diff
@@ -21,134 +21,100 @@ package org.apache.samza.system.kafka
                           metrics: KafkaSystemProducerMetrics,
-                          val clock: () => Long = () => System.nanoTime) extends SystemProducer with Logging with TimerUtils
-{
-
-  class SourceData {
-    /**
-     * lock to make send() and store its future atomic
-     */
-    val sendLock: Object = new Object
-    /**
-     * The most recent send's Future handle
-     */
-    @volatile
-    var latestFuture: Future[RecordMetadata] = null
-    /**
-     * exceptionInCallback: to store the exception in case of any "ultimate" send failure (ie. failure
-     * after exhausting max_retries in Kafka producer) in the I/O thread, we do not continue to queue up more send
-     * requests from the samza thread. It helps the samza thread identify if the failure happened in I/O thread or not.
-     *
-     * In cases of multiple exceptions in the callbacks, we keep the first one before throwing.
-     */
-    var exceptionInCallback: AtomicReference[SamzaException] = new AtomicReference[SamzaException]()
-  }
+                          val clock: () => Long = () => System.nanoTime,
+                          val dropProducerExceptions: Boolean = false) extends SystemProducer with Logging with TimerUtils {
 
+  // Represents a fatal error that caused the producer to close.
+  val fatalException: AtomicReference[SystemProducerException] = new AtomicReference[SystemProducerException]()
   @volatile var producer: Producer[Array[Byte], Array[Byte]] = null
-  var producerLock: Object = new Object
-  val StreamNameNullOrEmptyErrorMsg = "Stream Name should be specified in the stream configuration file."
-  val sources: ConcurrentHashMap[String, SourceData] = new ConcurrentHashMap[String, SourceData]
+  val producerLock: Object = new Object
 
   def start(): Unit = {
+    producer = getProducer()
   }
 
   def stop() {
-    try {
-      val currentProducer = producer
-      if (currentProducer != null) {
-        producerLock.synchronized {
-          if (currentProducer == producer) {
-            // only nullify the member producer if it is still the same object, no point nullifying new producer
-            producer = null
-          }
+    info(KafkaSystemProducer.StopMsg + this.systemName)
+
+    // stop() should not happen often so no need to optimize locking
+    producerLock.synchronized {
+      try {
+        if (producer != null) {
+          producer.close  // Also performs the equivalent of a flush()
         }
-        currentProducer.close
 
-        sources.asScala.foreach {p =>
-          if (p._2.exceptionInCallback.get() == null) {
-            flush(p._1)
-          }
+        // Log error for debugging purposes.
+        val exception = fatalException.get()
+        if (exception != null) {
+          error(KafkaSystemProducer.UnhandledSendErrorMsg, exception)
         }
+      } catch {
+        case e: Exception => error(KafkaSystemProducer.ErrorClosingMsg + systemName, e)
+      } finally {
+        producer = null
```

Comment:
This is the only place where producer is nullified. Might be clearer to update the null check message in send/flush to say something like "send/flush called after producer close".

Reply:
Could also happen if send/flush were called before start(). 

Either way, it's likely a samza producer lifecycle bug. I'm not sure getting more descriptive will help much.